File: Data Structures & Algorithms/network-delay-time/submission-4.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        // just bfs with a priority queue
        // hardest part is impl.
        vector<vector<pair<int, int>>> graph(n);
        for(int i = 0; i < times.size(); i++)
        {
            graph[times[i][0]-1].push_back(make_pair(times[i][1]-1, times[i][2]));
        }
        vector<int> vis(n, -1);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> bfs;
        bfs.push(make_pair(0, k-1));
        pair<int, int> curr;
        while(!bfs.empty())
        {
            curr = bfs.top();
            bfs.pop();
            if(vis[curr.second] != -1) continue;
            vis[curr.second] = curr.first;
            for(int i = 0; i < graph[curr.second].size(); i++)
            {
                if(vis[graph[curr.second][i].first] != -1) continue;
                bfs.push(make_pair(curr.first+graph[curr.second][i].second, graph[curr.second][i].first));
            }
        }
        int res = 0;
        for(int i = 0; i < n; i++)
        {
            if(vis[i] == -1) return -1;
            res = max(res, vis[i]);
        }
        return res;
    }
};

File: Data Structures & Algorithms/max-area-of-island/submission-4.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        # run bfs on grid
        stack = []
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                cnt = 0
                if grid[i][j] == 0:
                    continue
                stack.append((i, j))
                while len(stack) > 0:
                    x, y = stack[-1]
                    stack.pop()
                    grid[x][y] = 0
                    cnt = cnt + 1
                    for k in range(4):
                        nx = x + dx[k]
                        ny = y + dy[k]
                        if nx < 0 or nx >= len(grid) or ny < 0 or ny >= len(grid[0]):
                            continue
                        if grid[nx][ny] == 0:
                            continue
                        stack.append((nx, ny))
                        grid[nx][ny] = 0
                res = max(res, cnt)
        return res